src/workers/parser/mapOriginalExpression.js

```diff
@@ -81,22 +77,74 @@ export default function mapOriginalExpression(
     }
   }
 
-  if (replacements.size === 0) {
-    // Avoid the extra code generation work and also avoid potentially
-    // reformatting the user's code unnecessarily.
-    return expression;
+  return replacements;
+}
+
+// input `var a = 3` => `if (!window.hasOwnProperty("a")) window.a = 3 `
+function globalizeAssignment(node) {
+  return node.declarations.map(({ id, init }) => {
+    if (init) {
+      return t.ifStatement(
+        t.unaryExpression(
+          "!",
+          t.callExpression(
+            t.memberExpression(
+              t.identifier("self"),
+              t.identifier("hasOwnProperty")
+            ),
+            [t.stringLiteral(id.name)]
+          )
+        ),
+        t.expressionStatement(
+          t.assignmentExpression(
+            "=",
+            t.memberExpression(t.identifier("self"), id),
+            init
+          )
+        )
+      );
+    }
+  });
+}
+
+export default function mapOriginalExpression(
+  expression: string,
+  mappings: {
+    [string]: string | null
   }
+): string {
+  const ast = parseScript(expression);
+  const replacements = getReplacements(ast, mappings);
 
+  let didUpdate = false;
   t.traverse(ast, (node, ancestors) => {
-    if (!t.isIdentifier(node) && !t.isThisExpression(node)) {
+    const parent = ancestors[ancestors.length - 1];
+    if (!parent) {
       return;
     }
+    const parentNode = parent.node;
+
+    if (
+      replacements.size > 0 &&
+      (t.isIdentifier(node) || t.isThisExpression(node))
+    ) {
+      const replacement = replacements.get(locationKey(node.loc.start));
+      if (replacement) {
+        didUpdate = true;
+        replaceNode(ancestors, t.cloneNode(replacement));
+      }
+    }
+
+    if (t.isVariableDeclaration(node) && !t.isBlockStatement(parentNode)) {
+      const parts = globalizeAssignment(node);
 
-    const replacement = replacements.get(locationKey(node.loc.start));
-    if (replacement) {
-      replaceNode(ancestors, t.cloneNode(replacement));
+      didUpdate = true;
+      const lastAncestor = ancestors[ancestors.length - 1];
+      const { index } = lastAncestor;
+      parent.node[parent.key].splice(index, 1, ...parts);
```

Comment:
This will throw, since `if (true) var foo = 4` is a case where the parent is a single property, not an array of nodes.